### PR TITLE
(Smaller) PR to add MicrosoftGraphClientService to the new API.

### PR DESCRIPTION
**Pull request by jmprieur**
Fixes https://github.com/AzureAD/microsoft-identity-web/issues/427

Add the `MicrosoftGraphOptions`, `MicrosoftGraphServiceExtensions`, `TokenAcquisitionCredentialProvider` (coming from the Blazor sample)

Update the test apps that where using graph
- WebAppCallsMicrosoftGraph
- BlazorServerCallsGraph

Updating the samples requires the following:
- remove the package reference to the Microsoft Graph SDK, as it is no longer needed in the sample (brought by Microsoft.Identity.Web)
- update the appsettings.Json to use the new names for the settings (`BaseUrl` and `Scopes`)
- remove the implementation which was so far in the sample for the MicrosoftGraphService and it's token provider (this was the implementation brought by the project templates)
- in startup.cs, remove the calls to `services.AddMicrosoftGraph(scopes, Configuration.GetValue<string>("CalledApi:CalledApiUrl"));` and replace it by what we agreed on with Damien:

```CSharp
 .EnableTokenAcquisitionToCallDownstreamApi()
                            .AddMicrosoftGraphServiceClient(Configuration.GetSection("GraphBeta"))
```

### CODE REVIEW

File: src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 // Licensed under the MIT License.
 
-using Microsoft.Graph;
-using Microsoft.Identity.Web;
 using System.Collections.Generic;
 using System.Net.Http;
 using System.Threading.Tasks;
+using Microsoft.Graph;
 
-namespace blazor
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs:10]: Copied the TokenAcquisitionCredentialProvider from the blazor sample)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+using System;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Microsoft.Graph;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Extensions methods on a MicrososoftAppCallingWebApiAuthenticationBuilder builder
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:13]: nit: `MicrosoftIdentity...`

Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:13]:
```suggestion
    /// Extensions methods on a MicrososoftIdentityAppCallingWebApiAuthenticationBuilder builder
```

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    /// to add support to call Microsoft Graph.
+    /// </summary>
+    public static class MicrosoftGraphServiceExtensions
+    {
+        /// <summary>
+        /// Add support to calls Microsoft graph. From a named option and a configuration section.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:19]:
```suggestion
        /// Add support to call Microsoft Graph. From a named option and a configuration section.
```

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        public static MicrosoftIdentityAppCallsWebApiAuthenticationBuilder AddMicrosoftGraphServiceClient(
+            this MicrosoftIdentityAppCallsWebApiAuthenticationBuilder builder,
+            IConfigurationSection configurationSection)
+        {
+            return builder.AddMicrosoftGraphServiceClient(
+                options => configurationSection.Bind(options));
+        }
+
+        /// <summary>
+        /// Add support to calls Microsoft graph. From a base graph Url and a default scope.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:33]:
```suggestion
        /// Add support to call Microsoft Graph. From a base Graph Url and a default scope.
```

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            return builder.AddMicrosoftGraphServiceClient(
+                options =>
+                {
+                    options.BaseUrl = graphBaseUrl;
+                    options.Scopes = defaultScopes;
+                });
+        }
+
+        /// <summary>
+        /// Add support to calls Microsoft graph. From a named options and a configuraiton method.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:53]:
```suggestion
        /// Add support to call Microsoft Graph. From a named options and a configuration method.
```

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    /// Extensions methods on a MicrososoftAppCallingWebApiAuthenticationBuilder builder
+    /// to add support to call Microsoft Graph.
+    /// </summary>
+    public static class MicrosoftGraphServiceExtensions
+    {
+        /// <summary>
+        /// Add support to calls Microsoft graph. From a named option and a configuration section.
+        /// </summary>
+        /// <param name="builder">Builder.</param>
+        /// <param name="configurationSection">Configuraiton section.</param>
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:22]:
```suggestion
        /// <param name="configurationSection">Configuration section.</param>
```

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            return builder.AddMicrosoftGraphServiceClient(
+                options => configurationSection.Bind(options));
+        }
+
+        /// <summary>
+        /// Add support to calls Microsoft graph. From a base graph Url and a default scope.
+        /// </summary>
+        /// <param name="builder">Builder.</param>
+        /// <param name="graphBaseUrl">Named instance of option.</param>
+        /// <param name="defaultScopes">Configuraiton section.</param>
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:37]:
```suggestion
        /// <param name="defaultScopes">Configuration section.</param>
```

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            builder.Services.AddSingleton<GraphServiceClient, GraphServiceClient>(serviceProvider =>
+            {
+                var tokenAquisitionService = serviceProvider.GetRequiredService<ITokenAcquisition>();
+                var options = serviceProvider.GetRequiredService<IOptions<MicrosoftGraphOptions>>();
+
+                var microsoftGraphOptions = options.Value;
+                if (microsoftGraphOptions.Scopes == null)
+                {
+                    throw new ArgumentException("CalledApiScopes should not be null.");
+                }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:80]: this needs to move to the const file section for exceptions

**Comment by jennyf19** [comment]
> 

we'll have to write tests for this.

---
Refers to: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:94 in 14e1f35. [](commit_id = 14e1f35c23054739d3970b9d3ce7d1b8d6c578f3, deletion_comment = False)

File: src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs [hunk 9]
```diff
@@ -1,14 +1,16 @@
...
+namespace Microsoft.Identity.Web
 {
+    /// <summary>
+    /// Authentication provider based on MSAL.NET.
+    /// </summary>
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs:13]: nit: this comment seems a little misleading.

File: tests/WebAppCallsMicrosoftGraph/Startup.cs [hunk 10]
```diff
@@ -30,7 +30,8 @@ public void ConfigureServices(IServiceCollection services)
             services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
                     .AddMicrosoftIdentityWebApp(Configuration.GetSection("AzureAd"))
                         .EnableTokenAcquisitionToCallDownstreamApi()
-                           .AddInMemoryTokenCaches();  // Add a delegate overload. Should return the parent builder
+                           .AddMicrosoftGraphServiceClient(Configuration.GetSection("GraphBeta"))
```
Comment (jennyf19) [inline tests/WebAppCallsMicrosoftGraph/Startup.cs:33]: nit: question for asp .net maybe...`.AddMicrosoftGraph`? or leave as is...i'm not sure. `...ServiceClient` is more precise

**Comment by jennyf19** [comment]
looks good. i think there is some work to do around unit testing. i'll take see if i have time today to take a look at that.

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 5 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:53]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469486903](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469486903) [](ancestors = 469486903)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 7 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:37]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469487529](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469487529) [](ancestors = 469487529)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 4 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:33]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469486656](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469486656) [](ancestors = 469486656)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 6 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:22]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469487409](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469487409) [](ancestors = 469487409)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 3 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:19]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469486424](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469486424) [](ancestors = 469486424)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 2 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:13]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469486283](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469486283) [](ancestors = 469486283)

File: src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs (hunk 8 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/MicrosoftGraphServiceExtensions.cs:80]: fixed in this [PR](https://github.com/AzureAD/microsoft-identity-web/pull/442)

---
In reply to: [469488173](https://github.com/AzureAD/microsoft-identity-web/pull/434#discussion_r469488173) [](ancestors = 469488173)

File: src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs [hunk 11]
```diff
@@ -17,13 +19,19 @@ public TokenAcquisitionCredentialProvider(ITokenAcquisition tokenAcquisition, IE
...
+        /// </summary>
+        /// <param name="request">HttpRequest message to authenticate.</param>
+        /// <returns>A Task (as this is an async method).</returns>
         public async Task AuthenticateRequestAsync(HttpRequestMessage request)
         {
-            request.Headers.Add("Authorization",
-                $"Bearer {await _tokenAcquisition.GetAccessTokenForUserAsync(_initialScopes)}");
+            request.Headers.Add(
+                "Authorization",
+                $"Bearer {await _tokenAcquisition.GetAccessTokenForUserAsync(_initialScopes).ConfigureAwait(false)}");
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs:34]: nit: going to fix this to use constants as well, will be in a PR

Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs:34]: in same PR mentioned above.

File: src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs (hunk 9 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web/MicrosoftGraph/TokenAcquisitionCredentialProvider.cs:14]: Addressed

**Pull request merged**